### ISSUE TITLE
Memory copies between multiple devices

### ISSUE THREAD

**Issue by nmnobre** [AdaptiveCpp/AdaptiveCpp#1059]
Hi Aksel,

Last week, during a talk I gave on SYCL, someone in the audience asked about memory copies between devices.
They asked if the memory copies happened directly or via host memory.
I wasn't sure, but my gut told me that it'd be implementation dependent (as I don't think it's specified by the standard) but that hopefully it'd happen directly if possible.

Then I came across Lecture 13 of the SYCL Academy curriculum. There, it's stated that the copy will happen directly if the devices belong to the same context and via host memory otherwise.
Now, this is where confusion begins...
1) It seems that a context can only be associated with multiple devices on the same platform, but I think Open SYCL will happily build a context from devices on different platforms? If this is unintentional, it shouldn't be too hard to fix. :)
2) Open SYCL seems to place the two Quadro GV100 on my local workstation under the same platform (which I think is a good thing and would allow for them to also belong to the same context) but DPC++ does not, and DPC++ does prevent you from creating a context from devices on different platforms...

So... is there any implementation which allows for direct copies between devices that you are aware of?
I must confess I was a bit lazy and didn't dig into the code of either compiler.

Cheers,
-Nuno

**Comment by illuhad** [comment]
Hi Nuno,

I'm familiar with the SYCL academy material. It's worded a bit too strongly. It should say: *If two devices are not in the same context, the copy **may** go through host memory*

In the context of Open SYCL, forget context. It's the silliest thing and exists only because OpenCL has an object called "context" and SYCL 1.2.1 used to map directly from SYCL to OpenCL. Nobody knows what a SYCL context is supposed to be, and even less so what a SYCL context is supposed to be for a SYCL implementation with CUDA, HIP or OpenMP backends.

In Open SYCL, a context is nothing but a list of devices. The only thing we use it for is as a hint for USM memory management functions. If you call e.g. `malloc_shared()`, the implementation needs to know which backend to ask to perform the allocation. This is decided based on which devices are listed in the context.

Actual backend contexts (whatever this means) are managed internally by the runtime.

We do support contexts that span multiple platforms. This is part of the multi-device queue logic.

Open SYCL in principle supports direct copies between devices within one backend (except maybe with the Level Zero backend - but it has been developed against Intel iGPUs and is therefore untested in multi-device scenarios anyway).

For example, if you copy (e.g. using `handler::copy()`) between two USM device pointers that point to allocations on two different devices from the CUDA backend, it will emit a CUDA memcpy operation between the devices. That means that the exact same thing happens as if your application were using the CUDA runtime API directly.

However, currently we do not yet call the backend functions to explicitly enable P2P access. So depending on your system, it might be that CUDA or HIP internally route the memory copies through the host. If that is the case, adding that setup code should be an fix though :-)

It's a bit more complicated to express this in the buffer-accessor model. Note that the act of defining an accessor already causes data migration. So if you look at the following pseudo-code:

```c++
sycl::buffer b1;
sycl::buffer b2;
q.submit([&](handler& cgh){
  accessor a1{b1, cgh};
  accessor a2{b2, cgh};
  cgh.copy(a1, a2);
});
```

There will in general be three copies:
1.) Declaring a1 will cause the data in b1 to migrate to `q.get_device()`
2.) Declaring a2 will cause the data in b2 to migrate to `q.get_device()`
3.) Because accessors always refer to the device of `q.get_device()`,  `copy()` will ultimately always copy within the memory of a single device, not across devices.

Note that 1. and 2. will be optimized away if the most recent version of the buffer data is already present on device. Because accessor data migration ultimately relies on the same internal code paths as USM data copies, it will take advantage of backend capabilities to copy directly between devices.

For example, let's assume that we have two CUDA devices. Let's say b1 lives on the other CUDA device compared to `q.get_device()`, but b2 is already present on `q.get_device()`. Then two things will happen:
1. A single direct CUDA memcpy from the other device to `q.get_device()` for a1
2. The internal copy within the queue's device between a1 and b2 using another CUDA memcpy.

An important corollary from this is that the buffer's internal data migration logic will copy directly between devices if they are within one backend; so if you only have a single buffer and just access it on different devices, there will be no copies through the host on the SYCL level (but backends do internally what they do..)

So, tl;dr Open SYCL is probably your best shot at having a SYCL implementation that can directly copy between multiple devices :-) I'm again and again surprised by what does not seem to work in DPC++...

**Comment by nmnobre** [comment]
Thank you for the very detailed answer Aksel. 🤗

> In the context of Open SYCL, forget context. It's the silliest thing and exists only because OpenCL has an object called "context" and SYCL 1.2.1 used to map directly from SYCL to OpenCL. Nobody knows what a SYCL context is supposed to be, and even less so what a SYCL context is supposed to be for a SYCL implementation with CUDA, HIP or OpenMP backends.
> 

I can feel the rage through the screen. 😅

> For example, if you copy (e.g. using `handler::copy()`) between two USM device pointers that point to allocations on two different devices from the CUDA backend, it will emit a CUDA memcpy operation between the devices. That means that the exact same thing happens as if your application were using the CUDA runtime API directly.
> 

Perfect, I wasn't too far off then on my answer. 🙂

> There will in general be three copies: 1.) Declaring a1 will cause the data in b1 to migrate to `q.get_device()` 2.) Declaring a2 will cause the data in b2 to migrate to `q.get_device()` 3.) Because accessors always refer to the device of `q.get_device()`, `copy()` will ultimately always copy within the memory of a single device, not across devices.
> 
> Note that 1. and 2. will be optimized away if the most recent version of the buffer data is already present on device. Because accessor data migration ultimately relies on the same internal code paths as USM data copies, it will take advantage of backend capabilities to copy directly between devices.
> 

Agreed.

> For example, let's assume that we have two CUDA devices. Let's say b1 lives on the other CUDA device compared to `q.get_device()`, but b2 is already present on `q.get_device()`. Then two things will happen:
> 
> 1. A single direct CUDA memcpy from the other device to `q.get_device()` for a1
> 2. The internal copy within the queue's device between a1 and b2 using another CUDA memcpy.
> 
> An important corollary from this is that the buffer's internal data migration logic will copy directly between devices if they are within one backend; so if you only have a single buffer and just access it on different devices, there will be no copies through the host on the SYCL level (but backends do internally what they do..)

Okay, this all seems very reasonable and as expected, I'd say.
In your experience, is there any particular characteristic of the system which would impair P2P comms on say Nvidia GPUs?
Any easy way to test this other than timing/profiling the transfer?

> So, tl;dr Open SYCL is probably your best shot at having a SYCL implementation that can directly copy between multiple devices :-) I'm again and again surprised by what does not seem to work in DPC++...

Yeah, I mean, I don't mean to imply that DPC++ doesn't do the right thing.
As you said, I was probably reading too much from the strongly worded SYCL Academy material.
The two GPUs are definitely on different platforms though, and you can't create a context with them both, but maybe that doesn't mean anything.

**Comment by illuhad** [comment]
> Thank you for the very detailed answer Aksel.

Sure, no problem :)

> I can feel the rage through the screen. 

Apologies, I did not mean to rage ;) SYCL is starting to have some technical debts like every other model though ;)

> In your experience, is there any particular characteristic of the system which would impair P2P comms on say Nvidia GPUs?

I don't know for sure, but I could imagine that it's probably going to have an impact if the two GPUs are not connected to the same PCIe root complex. For example, if they are connected to separate CPUs in a dual-socket machine.

On CUDA you can obtain this information using `nvidia-smi topo`, for example `nvidia-smi topo -p2p r` to query P2P read capabilities or `nvidia-smi topo -mp` to print a matrix of affinities. But regarding the performance impact, there's probably no way but to benchmark.
As I said it's quite possible that we are currently missing a couple line of "for each device combination, enable P2P communication if possible" to take full advantage of P2P.

>  Yeah, I mean, I don't mean to imply that DPC++ doesn't do the right thing.
> As you said, I was probably reading too much from the strongly worded SYCL Academy material.
> The two GPUs are definitely on different platforms though, and you can't create a context with them both, but maybe > that doesn't mean anything.

I have some vague memory of someone telling me that multi-device with DPC++ and CUDA currently has issues because of this context/platform problem. So my null hypothesis is to believe that there's an issue :P
I think I've also heard that they do have a P2P extension for Intel hardware, but I don't know how it works.

**Comment by nmnobre** [comment]
That's all very useful information.
Thank you! :)
